langchain/src/stores/message/firestore.ts

```diff
@@ -0,0 +1,160 @@
+import * as firebaseAdmin from "firebase-admin";
+import { getApps, initializeApp } from "firebase-admin/app";
+
+import {
+  getFirestore,
+  DocumentData,
+  Firestore,
+  DocumentReference,
+  FieldValue,
+} from "firebase-admin/firestore";
+import {
+  StoredMessage,
+  BaseMessage,
+  BaseListChatMessageHistory,
+} from "../../schema/index.js";
+import {
+  mapChatMessagesToStoredMessages,
+  mapStoredMessagesToChatMessages,
+} from "./utils.js";
+
+export interface FirestoreDBChatMessageHistory {
+  collectionName: string;
+  sessionId: string;
+  userId: string;
+  appIdx?: number;
+  config?: firebaseAdmin.AppOptions;
+}
+export class FirestoreChatMessageHistory extends BaseListChatMessageHistory {
+  lc_namespace = ["langchain", "stores", "message", "firestore"];
+
+  private collectionName: string;
+
+  private sessionId: string;
+
+  private userId: string;
+
+  private appIdx: number;
+
+  private config: firebaseAdmin.AppOptions;
+
+  private firestoreClient: Firestore;
+
+  private document: DocumentReference<DocumentData> | null;
+
+  private messages: BaseMessage[];
+
+  constructor({
+    collectionName,
+    sessionId,
+    userId,
+    appIdx = 0,
+    config,
+  }: FirestoreDBChatMessageHistory) {
+    super();
+    this.collectionName = collectionName;
+    this.sessionId = sessionId;
+    this.userId = userId;
+    this.document = null;
+    this.appIdx = appIdx;
+    this.messages = [];
+    if (config) this.config = config;
+
+    try {
+      this.prepareFirestore();
+    } catch (error) {
+      throw new Error(`Unknown response type`);
+    }
+  }
+
+  private prepareFirestore(): void {
+    let app;
+    // Check if the app is already initialized else get appIdx
+    if (!getApps().length) app = initializeApp(this.config);
+    else app = getApps()[this.appIdx];
+
+    this.firestoreClient = getFirestore(app);
+
+    this.document = this.firestoreClient
```

Comment:
Needs to be awaited?

Reply:
No. they are not promises and references to firestore db and document

Reply:
Got it, thank you!